mentat/include_files.py

```diff
@@ -44,104 +23,224 @@ def is_file_text_encoded(abs_path: Path):
+                if any(include_patterns) and not match_path_with_patterns(
+                    abs_git_path, all_include_patterns
+                ):
+                    continue
+                if any(ignore_patterns) and match_path_with_patterns(
+                    abs_git_path, all_ignore_patterns
+                ):
+                    continue
+                paths.add(abs_git_path)
+
+        else:
+            filtered_dirs: List[str] = []
+            for dir_ in dirs:
+                abs_dir_path = root.joinpath(dir_)
+                if any(include_patterns) and not match_path_with_patterns(
+                    abs_dir_path, all_include_patterns
+                ):
+                    continue
+                if any(ignore_patterns) and match_path_with_patterns(
+                    abs_dir_path, all_ignore_patterns
+                ):
+                    continue
+                filtered_dirs.append(dir_)
+            dirs[:] = filtered_dirs
+
+            for file in files:
+                abs_file_path = root.joinpath(file)
+                if any(include_patterns) and not match_path_with_patterns(
+                    abs_file_path, all_include_patterns
+                ):
+                    continue
+                if any(ignore_patterns) and match_path_with_patterns(
+                    abs_file_path, all_ignore_patterns
+                ):
+                    continue
+                paths.add(abs_file_path)
+
+            if not recursive:
+                break
+
+    return paths
+
 
-    files: defaultdict[Path, list[CodeFeature]] = defaultdict(list)
-    for file in files_direct:
-        if file.path not in excluded_files:
-            files[file.path.resolve()].append(file)
+def get_code_features_for_path(
+    path: Path,
+    cwd: Path,
+    include_patterns: Iterable[Path | str] = [],
+    ignore_patterns: Iterable[Path | str] = [],
+) -> Set[CodeFeature]:
+    validated_path = validate_and_format_path(path, cwd)
+
+    # Directory
+    if validated_path.is_dir():
+        paths = get_paths_for_directory(
+            validated_path, include_patterns, ignore_patterns
+        )
+        code_features = set(CodeFeature(p) for p in paths)
+    # File
+    elif validated_path.is_file():
+        code_features = set([CodeFeature(validated_path)])
+    # File interval
+    elif len(str(validated_path).rsplit(":", 1)) > 1:
+        interval_path, interval_str = str(validated_path).rsplit(":", 1)
+        intervals = parse_intervals(interval_str)
+        code_features: Set[CodeFeature] = set()
+        for interval in intervals:
+            code_feature = CodeFeature(
+                f"{interval_path}:{interval.start}-{interval.end}"
+            )
+            code_features.add(code_feature)
+    # Glob pattern
+    else:
+        root_parts: List[str] = []
+        pattern: str | None = None
+        for i, part in enumerate(validated_path.parts):
+            if re.search(r"[\*\?\[\]]", str(part)):
```

Comment:
Again, why aren't we just using glob.glob?

Reply:
See https://github.com/AbanteAI/mentat/pull/286#issuecomment-1821996841

Reply:
See https://github.com/AbanteAI/mentat/pull/286#issuecomment-1821996841